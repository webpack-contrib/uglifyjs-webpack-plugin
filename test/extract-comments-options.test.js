import UglifyJsPlugin from '../src/index';
import { PluginEnvironment } from './helpers';

describe('when options.extractComments', () => {
  let compilation;
  let compilationEventBinding;

  beforeEach(() => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      uglifyOptions: {
        warnings: true,
        mangle: {
          properties: {
            builtins: true,
          },
        },
      },
      extractComments: true,
    });
    plugin.apply(compilerEnv);

    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();

    compilation = chunkPluginEnvironment.getEnvironmentStub();
    compilation.assets = {
      'test.js': {
        source: () => 'var foo = 1;',
      },
      'test1.js': {
        source: () =>
          'function foo(x) { if (x) { return bar(); not_called1(); } }',
        map: () => {
          return {
            version: 3,
            sources: ['test1.js'],
            names: ['foo', 'x', 'bar', 'not_called1'],
            mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
          };
        },
      },
    };
    compilation.warnings = [];
    compilation.errors = [];

    eventBinding.handler(compilation);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();
  });

  it('normalizes extractConmments', () => {
    compilationEventBinding.handler(
      [
        {
          files: ['test.js'],
        },
      ],
      () => {
        expect(compilation.errors.length).toBe(0);
      }
    );
  });

  it('outputs warnings for unreachable code', () => {
    compilationEventBinding.handler(
      [
        {
          files: ['test.js', 'test1.js'],
        },
      ],
      () => {
        expect(compilation.warnings.length).toBe(1);
        expect(compilation.warnings[0]).toBeInstanceOf(Error);
        expect(compilation.warnings[0].message).toEqual(
          expect.stringContaining('Dropping unreachable code')
        );
      }
    );
  });

  it('normalizes when options.extractComments is boolean', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      extractComments: true,
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation2 = chunkPluginEnvironment.getEnvironmentStub();
    compilation2.assets = {
      'test.js': {
        source: () => '// Comment\nvar foo = 1;',
      },
      'test1.js': {
        source: () => '/* Comment */\nvar foo = 1;',
      },
    };
    compilation2.warnings = [];
    compilation2.errors = [];

    eventBinding.handler(compilation2);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler(
      [
        {
          files: ['test.js', 'test1.js'],
        },
      ],
      () => {
        expect(compilation2.assets['test.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test1.js'].source()).toMatchSnapshot(
          'test1.js'
        );
        expect(
          compilation2.assets['test1.js.LICENSE'].source()
        ).toMatchSnapshot('test1.js.LICENSE');
        expect(compilation2.errors).toMatchSnapshot('errors');
        expect(compilation2.warnings).toMatchSnapshot('warnings');
      }
    );
  });

  it('normalizes when options.extractComments is regex', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      extractComments: /foo/,
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation2 = chunkPluginEnvironment.getEnvironmentStub();
    compilation2.assets = {
      'test.js': {
        source: () => '// Comment\nvar foo = 1;',
      },
      'test1.js': {
        source: () => '// foo\nvar foo = 1;',
      },
    };
    compilation2.warnings = [];
    compilation2.errors = [];

    eventBinding.handler(compilation2);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler(
      [
        {
          files: ['test.js', 'test1.js'],
        },
      ],
      () => {
        expect(compilation2.assets['test.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test1.js'].source()).toMatchSnapshot(
          'test1.js'
        );
        expect(
          compilation2.assets['test1.js.LICENSE'].source()
        ).toMatchSnapshot('test1.js.LICENSE');
        expect(compilation2.errors).toMatchSnapshot('errors');
        expect(compilation2.warnings).toMatchSnapshot('warnings');
      }
    );
  });

  it('normalizes when options.extractComments is string', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      extractComments: 'all',
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation2 = chunkPluginEnvironment.getEnvironmentStub();
    compilation2.assets = {
      'test.js': {
        source: () => '// Comment\nvar foo = 1;',
      },
      'test1.js': {
        source: () => '/* Comment */\nvar foo = 1;',
      },
    };
    compilation2.warnings = [];
    compilation2.errors = [];

    eventBinding.handler(compilation2);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler(
      [
        {
          files: ['test.js', 'test1.js'],
        },
      ],
      () => {
        expect(compilation2.assets['test.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test.js.LICENSE'].source()).toMatchSnapshot(
          'test.js.LICENSE'
        );
        expect(compilation2.assets['test1.js'].source()).toMatchSnapshot(
          'test1.js'
        );
        expect(
          compilation2.assets['test1.js.LICENSE'].source()
        ).toMatchSnapshot('test1.js.LICENSE');
        expect(compilation2.errors).toMatchSnapshot('errors');
        expect(compilation2.warnings).toMatchSnapshot('warnings');
      }
    );
  });

  it('normalizes when options.extractComments is function', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      extractComments: () => true,
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation2 = chunkPluginEnvironment.getEnvironmentStub();
    compilation2.assets = {
      'test.js': {
        source: () => '// Comment\nvar foo = 1;',
      },
      'test1.js': {
        source: () => '/* Comment */\nvar foo = 1;',
      },
    };
    compilation2.warnings = [];
    compilation2.errors = [];

    eventBinding.handler(compilation2);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler(
      [
        {
          files: ['test.js', 'test1.js'],
        },
      ],
      () => {
        expect(compilation2.assets['test.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test1.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test.js.LICENSE'].source()).toMatchSnapshot(
          'test.js.LICENSE'
        );
        expect(
          compilation2.assets['test1.js.LICENSE'].source()
        ).toMatchSnapshot('test1.js.LICENSE');
        expect(compilation2.errors).toMatchSnapshot('errors');
        expect(compilation2.warnings).toMatchSnapshot('warnings');
      }
    );
  });

  it('normalizes when options.extractComments is object', () => {
    const pluginEnvironment = new PluginEnvironment();
    const compilerEnv = pluginEnvironment.getEnvironmentStub();
    compilerEnv.context = '';

    const plugin = new UglifyJsPlugin({
      extractComments: {
        condition: true,
        filename(file) {
          return file.replace(/(\.\w+)$/, '.license$1');
        },
        banner(licenseFile) {
          return `License information can be found in ${licenseFile}`;
        },
      },
    });
    plugin.apply(compilerEnv);
    const [eventBinding] = pluginEnvironment.getEventBindings();
    const chunkPluginEnvironment = new PluginEnvironment();
    const compilation2 = chunkPluginEnvironment.getEnvironmentStub();
    compilation2.assets = {
      'test.js': {
        source: () => '// Comment\nvar foo = 1;',
      },
    };
    compilation2.warnings = [];
    compilation2.errors = [];

    eventBinding.handler(compilation2);
    [compilationEventBinding] = chunkPluginEnvironment.getEventBindings();

    compilationEventBinding.handler(
      [
        {
          files: ['test.js'],
        },
      ],
      () => {
        expect(compilation2.assets['test.js'].source()).toMatchSnapshot(
          'test.js'
        );
        expect(compilation2.assets['test.license.js'].source()).toMatchSnapshot(
          'test.license.js'
        );
        expect(compilation2.errors).toMatchSnapshot('errors');
        expect(compilation2.warnings).toMatchSnapshot('warnings');
      }
    );
  });
});
