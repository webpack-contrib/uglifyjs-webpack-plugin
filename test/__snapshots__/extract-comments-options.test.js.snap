// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when options.extractComments normalizes when options.extractComments is boolean: test.js 1`] = `"var foo=1;"`;

exports[`when options.extractComments normalizes when options.extractComments is boolean: test1.js 1`] = `"var foo=1;"`;

exports[`when options.extractComments normalizes when options.extractComments is function: errors 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is function: test.js 1`] = `
"/*! For license information please see test.js.LICENSE */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is function: test.js 2`] = `
"/*! For license information please see test1.js.LICENSE */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is function: test.js.LICENSE 1`] = `
"// Comment
"
`;

exports[`when options.extractComments normalizes when options.extractComments is function: test1.js.LICENSE 1`] = `
"/* Comment */
"
`;

exports[`when options.extractComments normalizes when options.extractComments is function: warnings 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is object: errors 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is object: test.js 1`] = `
"/*! License information can be found in test.license.js */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is object: test.license.js 1`] = `
"// Comment
"
`;

exports[`when options.extractComments normalizes when options.extractComments is object: warnings 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is regex: errors 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is regex: test.js 1`] = `"var foo=1;"`;

exports[`when options.extractComments normalizes when options.extractComments is regex: test1.js 1`] = `
"/*! For license information please see test1.js.LICENSE */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is regex: test1.js.LICENSE 1`] = `
"// foo
"
`;

exports[`when options.extractComments normalizes when options.extractComments is regex: warnings 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is string: errors 1`] = `Array []`;

exports[`when options.extractComments normalizes when options.extractComments is string: test.js 1`] = `
"/*! For license information please see test.js.LICENSE */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is string: test.js.LICENSE 1`] = `
"// Comment
"
`;

exports[`when options.extractComments normalizes when options.extractComments is string: test1.js 1`] = `
"/*! For license information please see test1.js.LICENSE */
var foo=1;"
`;

exports[`when options.extractComments normalizes when options.extractComments is string: test1.js.LICENSE 1`] = `
"/* Comment */
"
`;

exports[`when options.extractComments normalizes when options.extractComments is string: warnings 1`] = `Array []`;
